File: execute-upgrade.js
// Script para ejecutar upgrade de database
const { Pool } = require('pg');
const fs = require('fs');
const path = require('path');

// Configuración de conexión
const pool = new Pool({
    host: '34.66.86.207',
    port: 5433,
    database: 'n8n_db',
    user: 'n8n',
    password: 'Ef1g2025*'
});

async function executeUpgrade() {
    const client = await pool.connect();
    
    try {
        console.log('🔗 Conectado a PostgreSQL en Google Cloud');
        console.log('📂 Leyendo script SQL...\n');
        
        // Leer el script SQL
        const sqlScript = fs.readFileSync(path.join(__dirname, 'upgrade-database.sql'), 'utf8');
        
        console.log('⚡ Ejecutando script de upgrade...\n');
        console.log('=' .repeat(60));
        
        // Ejecutar el script completo
        await client.query(sqlScript);
        
        console.log('=' .repeat(60));
        console.log('\n✅ Script ejecutado exitosamente!\n');
        
        // Verificar tablas creadas
        console.log('📊 Verificando tablas creadas...\n');
        
        const result = await client.query(`
            SELECT table_name 
            FROM information_schema.tables 
            WHERE table_schema = 'public' 
              AND table_name IN ('auditoria', 'conversaciones', 'leads', 'notificaciones', 
                                 'promociones', 'uso_promociones', 'documentos', 
                                 'metricas_diarias', 'integraciones')
            ORDER BY table_name
        `);
        
        console.log('✅ Tablas nuevas creadas:');
        result.rows.forEach((row, i) => {
            console.log(`   ${i + 1}. ${row.table_name}`);
        });
        
        // Verificar triggers
        const triggers = await client.query(`
            SELECT trigger_name, event_object_table
            FROM information_schema.triggers
            WHERE trigger_schema = 'public'
            ORDER BY event_object_table, trigger_name
        `);
        
        console.log(`\n✅ Triggers creados: ${triggers.rows.length}`);
        triggers.rows.forEach((row, i) => {
            console.log(`   ${i + 1}. ${row.trigger_name} → ${row.event_object_table}`);
        });
        
        // Verificar vistas
        const views = await client.query(`
            SELECT table_name
            FROM information_schema.views
            WHERE table_schema = 'public'
            ORDER BY table_name
        `);
        
        console.log(`\n✅ Vistas creadas: ${views.rows.length}`);
        views.rows.forEach((row, i) => {
            console.log(`   ${i + 1}. ${row.table_name}`);
        });
        
        // Verificar columnas agregadas a clientes
        const clientesColumns = await client.query(`
            SELECT column_name
            FROM information_schema.columns
            WHERE table_name = 'clientes'
            ORDER BY ordinal_position
        `);
        
        console.log(`\n✅ Columnas en tabla 'clientes': ${clientesColumns.rows.length}`);
        
        // Verificar promociones insertadas
        const promociones = await client.query('SELECT codigo, descripcion FROM promociones');
        console.log(`\n✅ Promociones disponibles: ${promociones.rows.length}`);
        promociones.rows.forEach((row, i) => {
            console.log(`   ${i + 1}. ${row.codigo} - ${row.descripcion}`);
        });
        
        console.log('\n' + '='.repeat(60));
        console.log('🎉 DATABASE UPGRADE COMPLETADO EXITOSAMENTE! 🎉');
        console.log('='.repeat(60));
        
    } catch (error) {
        console.error('\n❌ Error ejecutando upgrade:', error.message);
        console.error('\n📍 Detalles del error:');
        console.error(error);
        process.exit(1);
    } finally {
        client.release();
        await pool.end();
    }
}

// Ejecutar
executeUpgrade().catch(err => {
    console.error('Error fatal:', err);
    process.exit(1);
});

File: execute-upgrade.py
import psycopg2
import sys

print("🔗 Conectando a PostgreSQL en Google Cloud...")

try:
    # Conectar a la base de datos
    conn = psycopg2.connect(
        host="34.66.86.207",
        port=5433,
        database="n8n_db",
        user="n8n",
        password="n8npass"
    )
    
    cur = conn.cursor()
    
    print("✅ Conexión establecida")
    print("📂 Leyendo script SQL...\n")
    
    # Leer el script SQL
    with open('upgrade-database.sql', 'r', encoding='utf-8') as f:
        sql_script = f.read()
    
    print("⚡ Ejecutando script de upgrade...")
    print("=" * 60)
    
    # Ejecutar el script
    cur.execute(sql_script)
    conn.commit()
    
    print("=" * 60)
    print("\n✅ Script ejecutado exitosamente!\n")
    
    # Verificar tablas creadas
    print("📊 Verificando tablas creadas...\n")
    
    cur.execute("""
        SELECT table_name 
        FROM information_schema.tables 
        WHERE table_schema = 'public' 
          AND table_name IN ('auditoria', 'conversaciones', 'leads', 'notificaciones', 
                             'promociones', 'uso_promociones', 'documentos', 
                             'metricas_diarias', 'integraciones')
        ORDER BY table_name
    """)
    
    tables = cur.fetchall()
    print(f"✅ Tablas nuevas creadas: {len(tables)}")
    for i, (table,) in enumerate(tables, 1):
        print(f"   {i}. {table}")
    
    # Verificar triggers
    cur.execute("""
        SELECT trigger_name, event_object_table
        FROM information_schema.triggers
        WHERE trigger_schema = 'public'
        ORDER BY event_object_table, trigger_name
    """)
    
    triggers = cur.fetchall()
    print(f"\n✅ Triggers creados: {len(triggers)}")
    for i, (trigger, table) in enumerate(triggers[:10], 1):  # Mostrar solo primeros 10
        print(f"   {i}. {trigger} → {table}")
    if len(triggers) > 10:
        print(f"   ... y {len(triggers) - 10} más")
    
    # Verificar vistas
    cur.execute("""
        SELECT table_name
        FROM information_schema.views
        WHERE table_schema = 'public'
        ORDER BY table_name
    """)
    
    views = cur.fetchall()
    print(f"\n✅ Vistas creadas: {len(views)}")
    for i, (view,) in enumerate(views, 1):
        print(f"   {i}. {view}")
    
    # Verificar promociones
    cur.execute("SELECT codigo, descripcion FROM promociones")
    promociones = cur.fetchall()
    print(f"\n✅ Promociones disponibles: {len(promociones)}")
    for i, (codigo, desc) in enumerate(promociones, 1):
        print(f"   {i}. {codigo} - {desc}")
    
    # Verificar columnas en clientes
    cur.execute("""
        SELECT column_name
        FROM information_schema.columns
        WHERE table_name = 'clientes'
        ORDER BY ordinal_position
    """)
    
    columns = cur.fetchall()
    print(f"\n✅ Columnas en tabla 'clientes': {len(columns)}")
    
    print("\n" + "=" * 60)
    print("🎉 DATABASE UPGRADE COMPLETADO EXITOSAMENTE! 🎉")
    print("=" * 60)
    
    cur.close()
    conn.close()
    
except Exception as e:
    print(f"\n❌ Error: {e}")
    sys.exit(1)
